release_notes/utils.py

```diff
@@ -0,0 +1,203 @@
+import collections
+import dataclasses
+import json
+import logging
+import time
+from typing import Optional
+
+import git
+import git.exc
+import github3.exceptions as gh3e
+import github3.pulls as gh3p
+import github3.structs as gh3s
+import semver
+import yaml
+import yaml.scanner
+from github3 import GitHub
+
+import release_notes.model as rnm
+
+_meta_key = 'gardener.cloud/release-notes-metadata/v1'
+logger = logging.getLogger(__name__)
+
+
+# pylint: disable=protected-access
+# noinspection PyProtectedMember
+def list_associated_pulls(gh: GitHub, owner: str, repo: str, sha: str) -> Optional[tuple[gh3p.ShortPullRequest]]:
+    ''' Returns a tuple with pull requests related to the specified commit.
+
+    :param gh: Instance of the GitHub v3 API
+    :param owner: Owner of the repository (on GitHub)
+    :param repo: Name of the repository (on GitHub)
+    :param sha: SHA of the commit
+    :return: a tuple with pull requests related to the specific commit
+    '''
+    try:
+        url = gh._build_url('repos', owner, repo, 'commits', sha, 'pulls')
+        return tuple(gh._iter(-1, url, gh3p.ShortPullRequest))
+    except gh3e.UnprocessableEntity as e:
+        return None
+
+
+def _write_to_git_notes(repo: git.Repo, commit: git.Commit, body: str):
+    ''' Notes can be attached to a commit using
+    `$ git notes add -m <message> <sha>`
+
+    :param repo: the repository the commit belongs to
+    :param commit: the commit to add notes to
+    :param body: the body to write to the commit notes
+    :return:
+    '''
+    print("writing\n", body, "\nto", commit.hexsha, "\n")
+    repo.git.notes('add', '-f', '-m', body, commit.hexsha)
+
+
+# pylint: disable=protected-access
+# noinspection PyProtectedMember
+def list_pulls(gh: GitHub, owner: str, repo: str, state: str = 'closed') -> gh3s.GitHubIterator[gh3p.ShortPullRequest]:
+    url = gh._build_url('repos', owner, repo, 'pulls') + '?state=' + state
+    return gh._iter(-1, url, gh3p.ShortPullRequest)
+
+
+def shorten(message: str, max_len: int = 128) -> str:
+    message = message.replace('\n', '\\n')
+    if len(message) > max_len:
+        message = message[:max_len - 3] + '...'
+    return message
+
+
+def create_release_notes_blocks(release_notes: set[rnm.ReleaseNote]) -> str:
+    return '\n\n'.join(z.block_str for z in release_notes)
+
+
+def find_next_smallest_version(available_versions: list[semver.VersionInfo],
+                               current_version: semver.VersionInfo) -> Optional[semver.VersionInfo]:
+    # find version before the requested version and sort by semver
+    return max((v for v in sorted(available_versions) if v < current_version), default=None)
+
+
+def _find_git_notes_for_commit(repo: git.Repo, commit: git.Commit) -> Optional[str]:
+    try:
+        return repo.git.notes('show', commit.hexsha)
+    except git.exc.GitCommandError:
+        return None
+
+
+def _find_payload_from_git_notes(repo: git.Repo, commit: git.Commit) -> Optional[dict]:
```

Comment:
is this function actually used/referenced?

Reply:
nope, it was a leftover - I've removed it.